src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,373 @@
+    size(std::move(layer.size)),
+    higherDimension(std::move(layer.higherDimension)),
+    runningMean(std::move(layer.runningMean)),
+    runningVariance(std::move(layer.runningVariance))
+{
+  // Nothing else to do.
+}
+
+template<typename MatType>
+BatchNormType<MatType>&
+BatchNormType<MatType>::operator=(const BatchNormType& layer)
+{
+  if (&layer != this)
+  {
+    Layer<MatType>::operator=(layer);
+    maxAxis = layer.maxAxis;
+    eps = layer.eps;
+    average = layer.average;
+    momentum = layer.momentum;
+    count = layer.count;
+    averageFactor = layer.averageFactor;
+    inputDimension = layer.inputDimension;
+    size = layer.size;
+    higherDimension = layer.higherDimension;
+    runningMean = layer.runningMean;
+    runningVariance = layer.runningVariance;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+BatchNormType<MatType>&
+BatchNormType<MatType>::operator=(
+    BatchNormType&& layer)
+{
+  if (&layer != this)
+  {
+    Layer<MatType>::operator=(std::move(layer));
+    maxAxis = std::move(layer.maxAxis);
+    eps = std::move(layer.eps);
+    average = std::move(layer.average);
+    momentum = std::move(layer.momentum);
+    count = std::move(layer.count);
+    averageFactor = std::move(layer.averageFactor);
+    inputDimension = std::move(layer.inputDimension);
+    size = std::move(layer.size);
+    higherDimension = std::move(layer.higherDimension);
+    runningMean = std::move(layer.runningMean);
+    runningVariance = std::move(layer.runningVariance);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::SetWeights(
+    typename MatType::elem_type* weightsPtr)
+{
+	MakeAlias(weights, weightsPtr, WeightSize(), 1);
+  // Gamma acts as the scaling parameters for the normalized output.
+	MakeAlias(gamma, weightsPtr, size, 1);
+  // Beta acts as the shifting parameters for the normalized output.
+	MakeAlias(beta, weightsPtr + gamma.n_elem, size, 1);
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::CustomInitialize(
+    MatType& W,
+    const size_t rows, 
+    const size_t /* cols */)
+{
+  if (rows != 2*size) {
+    throw std::invalid_argument("BatchNormType::CustomInitialize(): wrong "
+        "rows size!"); 
+  }
+  MatType gammaTemp;
+  MatType betaTemp;
+  // Gamma acts as the scaling parameters for the normalized output.
+  MakeAlias(gammaTemp, W.memptr(), size, 1);
+  // Beta acts as the shifting parameters for the normalized output.
+  MakeAlias(betaTemp, W.memptr() + gammaTemp.n_elem, size, 1);
+
+  gammaTemp.fill(1.0);
+  betaTemp.fill(0.0);
+
+  runningMean.zeros(size, 1);
+  runningVariance.ones(size, 1);
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Forward(
+    const MatType& input,
+    MatType& output)
+{
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  // Set size of output equal to the size of input.
+  // output.set_size(arma::size(input));
```

Comment:
I can't do multiline suggestions on a phone (or maybe I just haven't figured out how) but we can probably just remove these two lines, since `output`'s size should already be set right.